source/common/grpc/transcoding_filter.cc

```diff
@@ -0,0 +1,396 @@
+    }
+  }
+
+  return Http::FilterTrailersStatus::Continue;
+}
+
+void TranscodingFilter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+}
+
+Http::FilterHeadersStatus TranscodingFilter::encodeHeaders(Http::HeaderMap& headers,
+                                                           bool end_stream) {
+  log().debug("Transcoding::Instance::encodeHeaders {}", end_stream);
+  if (error_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (transcoder_) {
+    response_headers_ = &headers;
+    headers.insertContentType().value(Http::Headers::get().ContentTypeValues.Json);
+    if (!method_->server_streaming() && !end_stream) {
+      return Http::FilterHeadersStatus::StopIteration;
+    }
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus TranscodingFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  log().debug("Transcoding::Instance::encodeData");
+  if (error_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  if (transcoder_) {
+    response_in_.move(data);
+
+    if (end_stream) {
+      response_in_.finish();
+    }
+
+    ReadToBuffer(transcoder_->ResponseOutput(), data);
+
+    if (!method_->server_streaming()) {
+      return Http::FilterDataStatus::StopIterationAndBuffer;
+    }
+    // TODO(lizan): Check ResponseStatus
+  }
+
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus TranscodingFilter::encodeTrailers(Http::HeaderMap& trailers) {
+  log().debug("Transcoding::Instance::encodeTrailers");
+  if (transcoder_) {
+    response_in_.finish();
+
+    Buffer::OwnedImpl data;
+    ReadToBuffer(transcoder_->ResponseOutput(), data);
+
+    if (data.length()) {
+      encoder_callbacks_->addEncodedData(data);
+    }
+
+    if (!method_->server_streaming()) {
+      const Http::HeaderEntry* grpc_status_header = trailers.GrpcStatus();
+      if (grpc_status_header) {
+        uint64_t grpc_status_code;
+        if (!StringUtil::atoul(grpc_status_header->value().c_str(), grpc_status_code)) {
+          response_headers_->Status()->value(enumToInt(Http::Code::ServiceUnavailable));
+        }
+        response_headers_->insertGrpcStatus().value(*grpc_status_header);
+      }
+
+      const Http::HeaderEntry* grpc_message_header = trailers.GrpcMessage();
+      if (grpc_message_header) {
+        response_headers_->insertGrpcMessage().value(*grpc_message_header);
+      }
+
+      response_headers_->insertContentLength().value(
+          encoder_callbacks_->encodingBuffer() ? encoder_callbacks_->encodingBuffer()->length()
+                                               : 0);
+    }
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+void TranscodingFilter::setEncoderFilterCallbacks(Http::StreamEncoderFilterCallbacks& callbacks) {
+  encoder_callbacks_ = &callbacks;
+}
+
+bool TranscodingFilter::ReadToBuffer(google::protobuf::io::ZeroCopyInputStream* stream,
```

Comment:
@alyssawilk is working on adding watermarks to bound buffer sizes at the L4/L7 connection/stream management, but filters like this that buffer indefinitely are also an issue (this is also true for other filters we've added recently, e.g. grpc-web). I'd be interested in hearing from @mattklein123 and @alyssawilk on whether now is the time to start requiring filters to start (at least at config level) introducing the notion of bounded buffering.